### PR TITLE
feat: add monitor event-driven substrate

### LINKED ISSUE #63: [Feature] 监控与调试 (Monitoring & Debugging)
## 来源
参考 NVIDIA NemoClaw: https://github.com/NVIDIA/NemoClaw

## 功能描述
提供沙箱状态监控、日志流、网络活动监控的完整工具链。

## 核心能力
- 沙箱健康状态检查
- 实时日志流 (follow mode)
- TUI 终端界面监控网络活动
- 推理路由状态展示
- 被阻断请求可视化

## CLI 命令示例
```bash
# 检查沙箱状态
nemoclaw <name> status

# 查看日志
nemoclaw <name> logs
nemoclaw <name> logs -f

# TUI 监控
openshell term
```

## TUI 显示信息
- 活跃网络连接
- 待审批的网络请求
- 推理路由状态

## 优先级
中 - 运维可观测性

### PR DESCRIPTION
**Pull request by stevetdp**
## Summary
- add a reusable monitor/event-driven substrate with persisted subscriptions, trigger audit, cooldown and dedupe handling
- normalize background task output/state and GitHub webhook deliveries into monitor events, then route matches through the existing run wake-up/follow-up flow
- expose run-scoped monitor APIs/tools, add SSE/notification support, and document/test the new contract

## Validation
- `uv run --extra dev python -m compileall src/relay_teams tests/unit_tests`
- `uv run --extra dev ruff check src/relay_teams/monitors src/relay_teams/interfaces/server src/relay_teams/sessions/runs src/relay_teams/tools/workspace_tools src/relay_teams/notifications src/relay_teams/providers src/relay_teams/external_agents tests/unit_tests/monitors tests/unit_tests/interfaces/server/test_runs_router.py tests/unit_tests/sessions/runs/background_tasks/test_manager.py tests/unit_tests/notifications/test_notification_service.py tests/unit_tests/roles/test_builtin_role_tools.py tests/unit_tests/triggers tests/unit_tests/tools/registry/test_defaults.py`
- `uv run --extra dev ruff format --check src/relay_teams/monitors src/relay_teams/interfaces/server src/relay_teams/sessions/runs src/relay_teams/tools/workspace_tools src/relay_teams/notifications src/relay_teams/providers src/relay_teams/external_agents tests/unit_tests/monitors tests/unit_tests/interfaces/server/test_runs_router.py tests/unit_tests/sessions/runs/background_tasks/test_manager.py tests/unit_tests/notifications/test_notification_service.py tests/unit_tests/roles/test_builtin_role_tools.py tests/unit_tests/triggers tests/unit_tests/tools/registry/test_defaults.py`
- `uv run --extra dev basedpyright src/relay_teams/monitors src/relay_teams/interfaces/server src/relay_teams/sessions/runs src/relay_teams/tools/workspace_tools src/relay_teams/notifications src/relay_teams/providers src/relay_teams/external_agents tests/unit_tests/monitors tests/unit_tests/interfaces/server/test_runs_router.py tests/unit_tests/sessions/runs/background_tasks/test_manager.py tests/unit_tests/notifications/test_notification_service.py tests/unit_tests/roles/test_builtin_role_tools.py`
- `uv run --extra dev pytest -q tests/unit_tests/monitors/test_service.py tests/unit_tests/interfaces/server/test_runs_router.py tests/unit_tests/sessions/runs/background_tasks/test_manager.py tests/unit_tests/notifications/test_notification_service.py tests/unit_tests/roles/test_builtin_role_tools.py`
- `uv run --extra dev pytest -q tests/unit_tests/tools/registry/test_defaults.py tests/unit_tests/tools/test_package_data.py tests/unit_tests/interfaces/server/test_app.py tests/unit_tests/interfaces/server/test_container.py tests/unit_tests/providers/test_provider_factory.py tests/unit_tests/external_agents/test_provider.py tests/unit_tests/external_agents/test_host_tool_bridge.py tests/unit_tests/tools/workspace_tools/test_background_task_tools.py`
- `uv run --extra dev pytest -q tests/unit_tests`

Fixes #63

### CODE REVIEW

File: src/relay_teams/triggers/service.py [hunk 1]
```diff
@@ -0,0 +1,1887 @@
...
+            webhook_status=GitHubWebhookStatus.UNREGISTERED,
+            enabled=True,
+            created_at=now,
+            updated_at=now,
+        )
+        created = self._repository.create_repo_subscription(record)
+        if payload.register_webhook:
+            callback_url = _normalize_optional_text(payload.callback_url)
+            if callback_url is None:
+                raise ValueError("callback_url is required when register_webhook=true")
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py]: Fixed in `062b8628`.

`create_repo_subscription` now validates `callback_url` before persisting the repo subscription, so a failed `register_webhook=true` request no longer leaves behind a half-configured row. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_create_repo_subscription_requires_callback_before_persisting`.

File: src/relay_teams/triggers/repository.py [hunk 2]
```diff
@@ -0,0 +1,1258 @@
...
+    def list_repo_subscriptions_by_full_name(
+        self,
+        full_name: str,
+    ) -> tuple[GitHubRepoSubscriptionRecord, ...]:
+        rows = self._run_read(
+            lambda: self._conn.execute(
+                """
+                SELECT * FROM github_repo_subscriptions
+                WHERE full_name=?
+                ORDER BY created_at DESC
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/repository.py:517]: Fixed in `062b8628`.

Webhook lookup now passes `enabled_only=True` when resolving repo subscriptions by `repository.full_name`, so disabled repo subscriptions no longer participate in signature matching or trigger dispatch. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_handle_inbound_delivery_ignores_disabled_repo_subscriptions`.

File: src/relay_teams/triggers/service.py [hunk 3]
```diff
@@ -0,0 +1,1887 @@
...
+                "updated_at": _utc_now(),
+            }
+        )
+        persisted = self._repository.update_account(updated)
+        if "token" in payload.model_fields_set:
+            self._secret_store.set_token(
+                self._config_dir, account_id=account_id, token=token
+            )
+        if "webhook_secret" in payload.model_fields_set and webhook_secret is not None:
+            self._secret_store.set_webhook_secret(
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py]: Fixed in `062b8628`.

When `webhook_secret` is explicitly updated to `null`, the service now clears the secret store entry instead of leaving the previous secret active. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_update_account_clears_webhook_secret_when_explicitly_null`.

**Review by stevetdp — COMMENTED** [review]
## Code Review: Triggers + Monitors 系统实现

整体来看这是一个架构清晰的大功能 PR,核心实现了事件驱动的 Monitor 基板和 GitHub Webhook Trigger 管线。模块边界文档 `docs/system-module-boundaries.md` 写得很好。以下是具体意见:

---

### 架构亮点

- Monitor 基板与事件源解耦的设计正确,`MonitorEventEnvelope` 归一化后统一走 `MonitorService.emit()` 是好的抽象。
- `RunManager._route_system_message` 提取了 background task 通知和 monitor trigger 的公共路由逻辑,消除了重复代码。
- HMAC 签名验证、dedupe_key、cooldown 等安全与幂等机制完备。
- container.py 中统一用 `runtime.paths.config_dir` 替换传入的 `config_dir` 是正确的收拢。

---

### 问题与建议

#### 1. [Major] `triggers/service.py` 1893 行、`triggers/repository.py` 1263 行,单文件过大

建议拆分:
- `service.py` 可拆为 `account_service.py`、`subscription_service.py`、`delivery_service.py`(ingest+match)、`action_worker.py`(已独立但仍内联在 service 中)
- `repository.py` 可按聚合根拆为多个 repository 类

**优先级**: 建议 merge 前至少把 `GitHubTriggerActionWorker` 独立文件,它是后台线程逻辑,与主 service 混在一起增加维护成本。

#### 2. [Major] `delete_account` 不级联清理 repo subscriptions 和 webhooks

`triggers/service.py` `delete_account()` 删除 account 后,属于该 account 的 repo subscriptions、rules 和 GitHub webhook 都不会清理。会导致:
- GitHub 上的 webhook 成为孤儿,继续推送事件到已删除的 account
- Repo subscriptions 和 rules 变成孤儿数据

建议: 在 `delete_account` 中先 cascade 查找并删除所有关联的 rules 和 subscriptions,并在删除 subscription 前调用 unregister webhook。

同理,`delete_repo_subscription` 不 unregister webhook,可能导致远程 GitHub repo 上残留 webhook。

#### 3. [Major] `GitHubApiClient.list_pull_request_files` 没有分页

`triggers/github_client.py` 只做一次 GET 请求,GitHub API 每页默认 30 条,最多 300 个文件。大型 PR 会返回截断结果,导致 `paths_any` / `paths_ignore` 匹配不准。

建议: 加入分页循环,或至少加 `per_page=100` 参数并处理 `Link` header 分页。

#### 4. [Medium] `GitHubApiClient` 是纯同步的,在异步上下文中调用会阻塞事件循环

`github_client.py` 使用 `create_sync_http_client` (同步 httpx)。但 `ingest_github_webhook` 在 FastAPI 的 async handler 中被调用,其中 `_match_rules` 调用 `list_pull_request_files` 等会同步阻塞。高 webhook 并发时可能成为瓶颈。

建议: 添加 async 版本的 client 方法,或用 `asyncio.to_thread` 包装同步调用。

#### 5. [Medium] `llm_session.py` 中 `getattr` workaround

`llm_session.py:500`:
```python
monitor_service=cast(
    MonitorService | None,
    getattr(self, "_monitor_service", None),
),
```
这个 `getattr + cast` 是因为 `_monitor_service` 是在 `__init__` 中设置的但类型系统不可见吗? 如果 `__init__` 里已经有 `self._monitor_service = monitor_service`,应直接引用 `self._monitor_service`,不需要 getattr fallback。如果是向后兼容旧实例,应加 comment 说明原因。

#### 6. [Medium] `TriggerRepository` schema migration 在 `__init__` 中执行

每次实例化都跑 `_ensure_tables`,在高频创建场景(如测试)会有性能开销。更重要的是,checksum-based 的 migration检测如果在并发场景下初始化,可能导致竞态。

建议: 参考其他 repository 的模式,migration 应在应用启动时一次性执行,而非 repository 构造时。

#### 7. [Low] `container.py` trigger 相关服务未完整初始化

container 中创建了 `self.trigger_repository`、`self.github_api_client`、`self.github_trigger_secret_store` 但没有看到 `GitHubTriggerService` 的完整实例化(可能在 diff 截断部分)。如果确实缺失,则 trigger API router 会拿到 None 服务。请确认 `GitHubTriggerService` 已完整初始化并绑定到 router。

#### 8. [Low] `_BackgroundTaskRuntime.monitor_line_buffers` 和 `monitor_line_sequence` 无线程安全保护

`BackgroundTaskManager._emit_monitor_lines` 在 async IO callback 中修改 `runtime.monitor_line_sequence`,但如果同一个 runtime 的多个 stream 同时 produce output,dict 和 counter 的修改并非原子操作。目前因为 asyncio 是单线程协作,问题不大,但如果未来引入多线程,需要加锁。

#### 9. [Low] `MonitorActionType.EMIT_NOTIFICATION` 在部分代码中未被处理

`MonitorAction` 有 `emit_notification` action type,但 `_handle_monitor_trigger_local` 只显式处理了 `START_FOLLOWUP_RUN`、`WAKE_INSTANCE`、`WAKE_COORDINATOR`。如果 `EMIT_NOTIFICATION` 进来,会走 `_route_system_message` 逻辑,但 `allow_coordinator=False`,直接 fallback 到 `_spawn_system_followup_run`,这和"仅发通知"的语义可能不匹配。

#### 10. [Nit] container.py 中 `self.run_service._notification_service = self.notification_service` 直接写私有属性

```python
self.run_service._notification_service = self.notification_service
```
以及 `self.monitor_service.bind_action_sink(self.run_service)` 是命令式绑定。建议统一在构造函数中完成所有依赖注入,避免 post-init mutation。

---

### 总结

核心架构设计合理,模块边界清晰,建议重点处理:
1. `delete_account` / `delete_repo_subscription` 的级联清理和 webhook 注销
2. `list_pull_request_files` 分页支持
3. 单文件大小问题(尤其是 `service.py`)

File: src/relay_teams/triggers/service.py [hunk 4]
```diff
@@ -0,0 +1,1893 @@
...
+        normalized_payload = delivery.normalized_payload
+        existing = _json_string_tuple(normalized_payload.get("changed_files"))
+        if existing:
+            return existing
+        pull_request_number = _parse_int(
+            _json_text(normalized_payload.get("pull_request_number"))
+        )
+        if pull_request_number is None:
+            return ()
+        token = self._require_account_token(repo.account_id)
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py:1380]: Handled in `a6af26aa`. Path-based rule evaluation now treats missing account/system tokens the same as other changed-file lookup failures: `_match_rule` catches `ValueError` from `_require_account_token` and records a deterministic `changed_files_unavailable` non-match instead of aborting delivery processing. Added regression coverage in `tests/unit_tests/triggers/test_service.py`.

**Comment by liyanqi01** [comment]
Addressed the latest review follow-ups in `a6af26aa`:

- `delete_account` and `delete_repo_subscription` now do best-effort remote webhook cleanup before local cascade delete.
- `GitHubApiClient.list_pull_request_files` now paginates with `per_page=100`.
- GitHub delivery handling now runs in a threadpool so sync GitHub calls do not block the FastAPI event loop.
- Path-based rule evaluation now records `changed_files_unavailable` instead of aborting when the account token is missing.
- `GitHubTriggerActionWorker` was split into `src/relay_teams/triggers/action_worker.py` to reduce `service.py` sprawl.
- `AgentLlmSession` now repairs the final assistant message from the streamed text when the provider result text disagrees, which fixed the rolling-summary compaction CI flake.

Validation run locally:
- `uv run --extra dev ruff check --fix`
- `uv run --extra dev ruff format --no-cache --force-exclude --check`
- `uv run --extra dev basedpyright`
- `uv run --extra dev pytest -q tests/unit_tests`
- `uv run --extra dev pytest -q tests/integration_tests`
- `uv run --extra dev pytest -q tests/integration_tests/api/test_session_context_compaction.py::test_multiple_rolling_summary_rewrites_preserve_rounds_and_exact_recall` repeated 5x

**Comment by liyanqi01** [comment]
Follow-up on the CI-only compaction failure: the remaining issue was in the fake LLM harness, not the production compaction state machine. Rolling-summary compaction/recall responses were still being streamed in 12-character chunks, which reproduced an intermittent chunk-loss path only on GitHub Actions. `3c3e1455` keeps those responses streamed but sends them in large chunks (`chunk_size=4096`) so the test still exercises streaming without the flaky boundary behavior.

Revalidated locally after that change:
- `uv run --extra dev pytest -q tests/unit_tests`
- `uv run --extra dev pytest -q tests/integration_tests`
- rolling-summary exact-recall integration repeated 5x

**Review by stevetdp — COMMENTED** [review]
## Code Review: `feat: add monitor event-driven substrate`

Overall this is a well-structured feature addition. The monitor/trigger pipeline is cleanly separated into distinct modules with clear responsibilities. Below are specific findings.

---

### Architecture & Design

**Well done:**
- Clean separation of `monitors/` (subscription lifecycle, rule matching, dedupe/cooldown) from `triggers/` (GitHub webhook ingestion, external event routing).
- The `ActionSink` protocol in `MonitorService` avoids direct coupling to `RunManager` — good inversion of control.
- `delete_account` in `TriggerRepository` (line 330) properly cascades through all child tables in dependency order (action_attempts → dispatches → evaluations → rules → subscriptions → accounts).

**Concerns:**

1. **`trigger_service.py` is 1,877 lines.** This is significantly above the project norm. Consider extracting the GitHub webhook ingestion pipeline (delivery/signature verification, rule evaluation, action dispatch) into a dedicated module (`github_webhook_pipeline.py` or similar). The `GitHubTriggerService` class has at least 5 distinct responsibility clusters: account CRUD, subscription CRUD, rule CRUD, webhook ingestion/matching, and the action worker loop.

2. **`run_manager.py` grows by +238/−52 lines.** The `_handle_monitor_trigger_local` and `_route_system_message` methods add monitor-specific routing to an already large file. Consider routing monitor actions through `MonitorService._dispatch_action` exclusively (it already handles `EMIT_NOTIFICATION` locally) and letting the action sink pattern handle all action types — this would keep `RunManager` as a thin sink rather than a router.

---

### Specific Code Issues

3. **`container.py:707` — private attribute access:**
   ```python
   self.run_service._notification_service = self.notification_service
   ```
   This reaches into `run_service` internals. If `RunService` needs a notification service, inject it via constructor or a setter method rather than directly assigning a private attribute from outside.

4. **`container.py:596` — late binding pattern:**
   ```python
   self.monitor_service.bind_action_sink(self.run_service)
   ```
   Combined with `bind_notification_service` at line 708, this two-phase initialization is fragile. The monitor service is partially usable between construction and binding — any event that arrives during that window will silently drop actions (`_action_sink is None` check at `monitor_service.py:244`). Consider providing all collaborators at construction time, or at minimum adding a startup guard.

5. **`trigger_repo.py:330-357` — raw SQL DELETE without transaction safety:**
   The `delete_account` method executes 6 sequential DELETEs inside `_run_write`. If these aren't wrapped in a single transaction, a partial failure could leave orphaned rows. Verify that `_run_write` wraps the entire lambda in a single transaction.

6. **`github_client.py:95-130` — unbounded pagination:**
   `list_pull_request_files` paginates in a `while True` loop with no upper bound on page count. A PR with thousands of files could stall the worker. Add a `max_pages` or `max_files` safety limit.

7. **`trigger_service.py:449-450` — `delete_rule` does not cascade to dependent triggers:**
   ```python
   def delete_rule(self, trigger_rule_id: str) -> None:
       _ = self._repository.get_rule(trigger_rule_id)
       self._repository.delete_rule(trigger_rule_id)
   ```
   Unlike `delete_account` which cascades through child tables, `delete_rule` may orphan dispatches and action_attempts that reference this rule. Verify the repo-level DELETE has CASCADE or explicit child cleanup.

---

### Minor / Style

8. **`trigger_service.py:227-237` — unused variable:**
   ```python
   _ = self._repository.get_account(account_id)
   ```
   The `_` assignment is used as an existence check. Consider a dedicated `account_exists` query or at least rename to `_account` for clarity — the `_` convention is easy to miss during review.

9. **`build_trigger_message` in `monitor_service.py:222-237`** builds XML-like strings via f-string interpolation. This is fine for now, but `_xml_escape` must cover all edge cases (CDATA delimiters, null bytes). Ensure test coverage for adversarial input (e.g. `body_text` containing `</monitor-event>`).

---

### Testing

10. Unit test coverage for the new modules looks good (`test_service.py` for both monitors and triggers). I'd also suggest:
    - **Integration test for the webhook delivery → trigger evaluation → monitor dispatch → run wake-up** end-to-end path.
    - **Concurrency test**: two webhooks for the same subscription arriving simultaneously should test dedupe correctness under race conditions.

---

### Summary

Requesting changes primarily on:
- Extracting `trigger_service.py` into smaller, focused modules (item 1)
- Replacing `run_service._notification_service = ...` with proper DI (item 3)
- Adding a safety bound to `list_pull_request_files` pagination (item 6)
- Verifying `delete_rule` cascade semantics (item 7)

File: src/relay_teams/triggers/service.py [hunk 5]
```diff
@@ -0,0 +1,1887 @@
...
+                        "attempt_count": attempt.attempt_count + 1,
+                        "last_error": "delivery is missing repo_subscription_id",
+                        "updated_at": _utc_now(),
+                    }
+                )
+            )
+            _ = updated
+            return True
+        repo = self._repository.get_repo_subscription(repo_subscription_id)
+        token = self._require_account_token(repo.account_id)
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py]: Handled in `7a961390`.

`_process_attempt` now resolves the GitHub token inside its local `try/except`, so a missing account/system token marks only that action attempt as failed instead of aborting `process_pending_actions()` and blocking later attempts. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_process_pending_actions_marks_attempt_failed_when_token_missing`.

File: src/relay_teams/triggers/service.py [hunk 6]
```diff
@@ -0,0 +1,1887 @@
...
+            owner=owner,
+            repo=repo_name,
+        )
+        now = _utc_now()
+        record = GitHubRepoSubscriptionRecord(
+            repo_subscription_id=f"ghrs_{uuid.uuid4().hex[:12]}",
+            account_id=account.account_id,
+            owner=owner,
+            repo_name=repo_name,
+            full_name=f"{owner}/{repo_name}",
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py]: Handled in `7a961390`.

`create_repo_subscription` now derives `owner`, `repo_name`, and `full_name` from GitHub’s canonical repository payload before persisting the subscription, so webhook matching no longer depends on the caller’s input casing. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_create_repo_subscription_uses_canonical_repository_identity`.

File: src/relay_teams/triggers/service.py [hunk 7]
```diff
@@ -0,0 +1,1887 @@
...
+            normalized_headers.get(_GITHUB_EVENT_HEADER)
+        )
+        signature = _normalize_optional_text(
+            normalized_headers.get(_GITHUB_SIGNATURE_HEADER)
+        )
+        if provider_delivery_id is not None:
+            existing = self._repository.get_delivery_by_provider_id(
+                provider=TriggerProvider.GITHUB.value,
+                provider_delivery_id=provider_delivery_id,
+            )
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py:544]: Handled in `7a961390`.

`TriggerRepository.create_delivery` now maps `(provider, provider_delivery_id)` unique-index collisions to a delivery conflict, and `handle_inbound_github_delivery` converts that race into the same duplicate response path as the normal pre-read. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_handle_inbound_delivery_returns_duplicate_on_insert_race`.

File: src/relay_teams/triggers/service.py [hunk 8]
```diff
@@ -0,0 +1,1893 @@
...
+            return {
+                "trigger_delivery_id": delivery.trigger_delivery_id,
+                "ingest_status": delivery.ingest_status.value,
+                "dispatch_count": 0,
+            }
+        subscriptions = self._repository.list_repo_subscriptions_by_full_name(
+            repository_full_name,
+            enabled_only=True,
+        )
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/triggers/service.py:612]: Handled in `7a961390`.

Inbound GitHub deliveries now filter repo subscriptions through enabled trigger accounts before signature matching and rule evaluation, so disabling an account fully suppresses webhook-driven dispatches for that account. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_handle_inbound_delivery_ignores_disabled_trigger_accounts`.

File: src/relay_teams/sessions/runs/run_manager.py [hunk 9]
```diff
@@ -1686,6 +1812,50 @@ def stop_subagent(self, run_id: str, instance_id: str) -> dict[str, str]:
...
+        created_by_instance_id: str | None = None,
+        created_by_role_id: str | None = None,
+        tool_call_id: str | None = None,
+    ) -> dict[str, object]:
+        service = self._require_monitor_service()
+        record = service.create_monitor(
+            run_id=run_id,
+            session_id=self._require_run_session_id(run_id),
+            source_kind=source_kind,
+            source_key=source_key,
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/sessions/runs/run_manager.py]: Handled in `7a961390`.

`RunManager.create_monitor` now validates `background_task` sources with `get_background_task(run_id=..., background_task_id=...)` before persisting the monitor, so a run cannot subscribe to another run’s background task id. Added regression coverage in `tests/unit_tests/sessions/runs/test_run_manager_recovery.py::test_create_monitor_validates_background_task_belongs_to_run`.

File: src/relay_teams/monitors/service.py [hunk 10]
```diff
@@ -0,0 +1,363 @@
...
+            source_kind=envelope.source_kind.value,
+            source_key=envelope.source_key,
+        )
+        for subscription in subscriptions:
+            if not _rule_matches(subscription.rule, envelope):
+                continue
+            if envelope.dedupe_key and self._repository.has_trigger_dedupe_key(
+                monitor_id=subscription.monitor_id,
+                dedupe_key=envelope.dedupe_key,
+            ):
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/monitors/service.py]: Handled in `7a961390`.

Monitor trigger recording now goes through `MonitorRepository.record_matching_trigger`, which performs the dedupe lookup, cooldown/max-trigger check, subscription update, and trigger insert inside one sqlite write section. That removes the stale read/update window that could double-fire the same monitor event. The existing monitor service regression coverage (`tests/unit_tests/monitors/test_service.py`) now exercises the new atomic path.

**Comment by liyanqi01** [comment]
Addressed the remaining review follow-ups in `7a961390`:

- disabled trigger accounts are now filtered out during inbound delivery matching
- missing action tokens now fail only the individual action attempt instead of aborting the worker loop
- repo subscriptions persist GitHub’s canonical repository identity
- delivery insert races on `provider_delivery_id` now resolve to the duplicate-delivery path
- run-scoped monitor creation now validates background-task ownership before persisting
- monitor dedupe/cooldown/max-trigger updates now happen through one atomic repository path

Revalidated locally after this batch:
- `uv run --extra dev ruff check --fix`
- `uv run --extra dev ruff format --no-cache --force-exclude --check`
- `uv run --extra dev basedpyright`
- `uv run --extra dev pytest -q tests/unit_tests`
- `uv run --extra dev pytest -q tests/integration_tests`

File: src/relay_teams/agents/execution/conversation_compaction.py [hunk 11]
```diff
@@ -561,15 +630,30 @@ def _render_transcript(
...
+        return ("", False)
+    if max_chars <= 0:
+        return ("", True)
+    if len(stripped) <= max_chars:
+        return (stripped, False)
+    prefix = stripped[:max_chars]
+    last_newline = prefix.rfind("\n")
+    if last_newline <= 0:
+        return ("", True)
+    return (prefix[:last_newline].rstrip(), True)
```
[the remark this answers is not in the document]
Comment (liyanqi01) [reply src/relay_teams/agents/execution/conversation_compaction.py]: Handled in `b295c9f7`.

`_clip_rendered_message` now falls back to safe boundaries instead of arbitrary character cuts: it still snaps to the last complete newline when later body lines are available, and otherwise trims back to the last safe whitespace boundary in the first body line. That preserves meaningful content for simple single-line messages without reintroducing partial fact tokens like `lunar-min` from long shell output lines. Added regression coverage in `tests/unit_tests/agents/execution/test_conversation_compaction.py::test_render_transcript_keeps_single_line_message_body_when_truncated` and `tests/unit_tests/agents/execution/test_conversation_compaction.py::test_render_transcript_does_not_clip_first_tool_line_mid_token`, then revalidated the rolling-summary exact-recall integration path repeatedly.

**Comment by liyanqi01** [comment]
Follow-up on the latest compaction review in `a0ec6bd6`: transcript clipping now preserves truncated single-line message bodies instead of dropping back to the role header only, while still snapping to full-line boundaries once later body lines are present.

Revalidated locally after that fix:
- `uv run --extra dev pytest -q tests/unit_tests/agents/execution/test_conversation_compaction.py tests/integration_tests/api/test_session_context_compaction.py::test_multiple_rolling_summary_rewrites_preserve_rounds_and_exact_recall`
- rolling-summary exact-recall integration repeated 5x
- `uv run --extra dev pytest -q tests/unit_tests`
- `uv run --extra dev pytest -q tests/integration_tests`

**Comment by liyanqi01** [comment]
Final follow-up in `b295c9f7`: compaction transcript clipping now trims at safe boundaries instead of preserving arbitrary partial tokens from long first-line tool output. That keeps single-line body content available for the summary rewrite while avoiding corrupted exact facts like `lunar-min`.

Revalidated locally after this adjustment:
- `uv run --extra dev pytest -q tests/unit_tests/agents/execution/test_conversation_compaction.py tests/integration_tests/api/test_session_context_compaction.py::test_multiple_rolling_summary_rewrites_preserve_rounds_and_exact_recall`
- rolling-summary exact-recall integration repeated 5x
- `uv run --extra dev pytest -q tests/unit_tests`
- `uv run --extra dev pytest -q tests/integration_tests`

File: src/relay_teams/monitors/repository.py [hunk 12]
```diff
@@ -0,0 +1,541 @@
...
+
+    def _init_tables(self) -> None:
+        def operation() -> None:
+            self._conn.execute(
+                """
+                CREATE TABLE IF NOT EXISTS monitor_subscriptions (
+                    monitor_id TEXT PRIMARY KEY,
+                    run_id TEXT NOT NULL,
+                    session_id TEXT NOT NULL,
+                    source_kind TEXT NOT NULL,
```
[the remark this answers is not in the document]
Comment (stevetdp) [reply src/relay_teams/monitors/repository.py:34]: Fixed in e1480cb2. `SessionService.delete_session()` now calls `MonitorRepository.delete_by_session(session_id)` via the server container, so monitor subscriptions and trigger records are removed alongside the session runtime/background-task cleanup path. I also added repository helpers for run/session-scoped monitor cleanup and extended `tests/unit_tests/sessions/test_session_delete_workspace_cleanup.py` to cover the cascade.

File: src/relay_teams/triggers/service.py [hunk 13]
```diff
@@ -0,0 +1,1967 @@
...
+            return (
+                False,
+                "check_conclusion_mismatch",
+                "check_conclusion did not match",
+            )
+        if match_config.paths_any or match_config.paths_ignore:
+            try:
+                changed_files = self._resolve_changed_files(
+                    delivery=delivery, repo=repo
+                )
```
[the remark this answers is not in the document]
Comment (stevetdp) [reply src/relay_teams/triggers/service.py]: Fixed in 7988a215. Changed-files resolution is now cached per delivery: `_evaluate_delivery()` creates a lazy cache and passes it through `_match_rule()`, so multiple path-filtered rules share one `list_pull_request_files` lookup and the same success/failure result. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_handle_inbound_delivery_caches_changed_files_lookup_per_delivery`.

File: src/relay_teams/triggers/service.py [hunk 14]
```diff
@@ -0,0 +1,1967 @@
...
+                continue
+            dispatches.append(
+                self._create_dispatch(delivery=delivery, rule=rule, repo=repo)
+            )
+            self._repository.update_rule(
+                rule.model_copy(
+                    update={
+                        "last_fired_at": _utc_now(),
+                        "last_error": None,
+                        "updated_at": _utc_now(),
```
[the remark this answers is not in the document]
Comment (stevetdp) [reply src/relay_teams/triggers/service.py:758]: Fixed in 7988a215. `_evaluate_delivery()` now only clears `rule.last_error` when the matched dispatch actually starts cleanly; if `_create_dispatch()` returns a failed dispatch, the rule keeps that dispatch error so operators can still see the startup failure. Added regression coverage in `tests/unit_tests/triggers/test_service.py::test_handle_inbound_delivery_preserves_rule_last_error_when_dispatch_fails`.

**Review by liyanqi01 — APPROVED** [review]
